File: samples/java/ZoneWorld/Client/src/main/java/systems/zlink/samples/zoneworld/client/ClientOptions.java
package systems.zlink.samples.zoneworld.client;

import java.io.IOException;
import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Objects;
import java.util.Properties;

public record ClientOptions(
        String gatewayEndpoint,
        String opsEndpoint,
        String scenarios,
        boolean streamTrace,
        String faultArmFile) {
    public static ClientOptions load(String[] args) throws IOException {
        if (args.length != 2 || !"--config".equals(args[0]) || args[1].isBlank()) {
            throw new IllegalArgumentException("Usage: ZoneWorldClient --config <path>");
        }
        Properties properties = new Properties();
        try (Reader reader = Files.newBufferedReader(Path.of(args[1]))) {
            properties.load(reader);
        }
        return new ClientOptions(
                require(properties, "sample.gateway-endpoint"),
                require(properties, "sample.ops-endpoint"),
                properties.getProperty("sample.scenarios", "all"),
                Boolean.parseBoolean(properties.getProperty("sample.stream-trace", "false")),
                properties.getProperty("sample.fault-arm-file", ""));
    }

    private static String require(Properties properties, String key) {
        String value = Objects.requireNonNullElse(properties.getProperty(key), "");
        if (value.isBlank()) throw new IllegalArgumentException(key + " is required");
        return value;
    }
}

File: samples/kotlin/ZoneWorld/Client/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/client/ClientOptions.kt
package systems.zlink.samples.kotlin.zoneworld.client

import java.io.Reader
import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties

data class ClientOptions(
    val gatewayEndpoint: String,
    val opsEndpoint: String,
    val scenarios: String,
    val streamTrace: Boolean,
    val faultArmFile: String,
) {
    companion object {
        fun load(args: Array<String>): ClientOptions {
            require(args.size == 2 && args[0] == "--config" && args[1].isNotBlank()) {
                "Usage: ZoneWorldClient --config <path>"
            }
            val properties = Properties()
            Files.newBufferedReader(Path.of(args[1])).use { reader: Reader ->
                properties.load(reader)
            }
            return ClientOptions(
                gatewayEndpoint = required(properties, "sample.gateway-endpoint"),
                opsEndpoint = required(properties, "sample.ops-endpoint"),
                scenarios = properties.getProperty("sample.scenarios", "all"),
                streamTrace = properties.getProperty("sample.stream-trace", "false").toBoolean(),
                faultArmFile = properties.getProperty("sample.fault-arm-file", ""),
            )
        }

        private fun required(properties: Properties, key: String): String =
            properties.getProperty(key)?.takeIf { it.isNotBlank() } ?: error("$key is required")
    }
}
